src/_P098_CacheReader.ino

```diff
@@ -0,0 +1,478 @@
+  Sample_t *sample;
+  byte *buffer;
+  float *fileNr;
+  float *offset;
+  float *sampleCount;
+  float *sampleIndex;
+  bool loadFromServer;
+  bool fileOpen; 
+  Cache_t(){
+    // Create Sample
+    sample = new Sample_t();
+  }
+  ~Cache_t(){
+    //delete &cache
+    delete sample;
+    delete[] buffer;
+  }
+  void initialize(struct EventStruct *event){
+    // Load values from UserVar array
+    fileNr = &UserVar[event->BaseVarIndex];
+    offset = &UserVar[event->BaseVarIndex +1];
+    sampleCount = &UserVar[event->BaseVarIndex +2];
+    sampleIndex = &UserVar[event->BaseVarIndex +3];
+    
+    // Initialize buffer for sample data
+    buffer = new byte[24];
+    for (int i=0; i<24; i++){
+      buffer[i] = '\0';
+    }
+    if (*fileNr < 1){
+      this->fetchFileNr();
+    }
+    if (*sampleIndex < 1){
+      this->fetchSampleIndex();
+    }
+    if (*offset < 1){
+      this->getOffset();
+    }
+    this->setData();
+  }
+  void fetchFileNr(){
+    // TODO: fetch file nr from server
+    (*this->fileNr) = 1;
+  }
+  void fetchSampleIndex(){
+    // TODO: fetch current sample index from server
+  }
+  void getOffset(){
+    // TODO: Find offset instead of static 16 byte offset
+    float staticOffset = 16;
+    memcpy(offset, &staticOffset, 4);
+  }
+  bool setData(){
+    /*
+    char *filename = new char[24];
+    for (int i = 0 ; i<24; i++){
+      filename[i] = '\0';
+    }
+    sprintf(filename,"cache_%d.bin",fnr);
+    */
+    int fnr = (int)(*this->fileNr);
+    char *filename = returnFilename(fnr);
+
+
+    if (!fileExists(filename)){
+      this->nextFile();
+      return false;
+    }
+    cache = tryOpenFile(filename,"r");
+    fileOpen = true;
+
+    delete[] filename;
+    return true;
+  }
+  void readSample(struct EventStruct *event){
+    if (fileOpen){
+    // Seek to current sample index
+    // TODO: Check return value if seek is out of bounds!
+    cache.seek(24 * (*sampleIndex) + (*offset));
+    // Read sample data into buffer
+    size_t read = cache.read(buffer, 24);
+    // Check if read was valid and does not contain EOF
+    if (read == 24 && buffer[23] != EOF){
+      // Parse buffer into Sample_t
+      sample->parseSample(buffer);
+      // Validate and send sample
+      bool validated = sample->validateSample();
+      if (validated){
+        sample->sendSample(event);
+      }
+      // Increment to next sample
+      this->increment();
+    } else {
+      this->deleteFile();
+      this->nextFile();
+    }
+    }else {
+      this->nextFile();
+    }
+  }
+  void setIndex(int index){
+    // TODO:
+  }
+  void increment(){
+    // TODO: Check if last sample in file
+    // If so, move to next file
+    (*this->sampleIndex)++;
+  }
+  void deleteFile(){
+    /*
+    int fnr = (int)(*this->fileNr);
+    char *filename = new char[24];
+    for (int i = 0 ; i<24; i++){
+      filename[i] = '\0';
+    }
+    sprintf(filename,"cache_%d.bin",fnr);
+    */
+    int fnr = (int)(*this->fileNr);
```

Comment:
You can assign a float to an int, without casting it.
However, you may run into rounding errors, so perhaps this needs some extra checks.